pkg/kubelet/volumemanager/reconciler/reconciler.go

```diff
@@ -251,77 +239,287 @@ func (rc *reconciler) reconciliationLoopFunc() func() {
 			}
 		}
 	}
 }
+
+func (rc *reconciler) reconstructFunc() {
+	glog.V(1).Infof("Start the routine to reconstruct volumes from disk directories %q", rc.kubeletPodsDir)
+	rc.reconstructFromDisk(rc.kubeletPodsDir)
+	rc.timeOfLastReconstruct = time.Now()
+}
+
+type podVolume struct {
+	podName        string
+	volumeSpecName string
+	mountPath      string
+	pluginName     string
+}
+
+// reconstructFromDisk reads the volume directories under the given pod directory. If the volume is not currently
+// in either actual or desired state of world, or in pending operation, reconstructFromDisk function will reconstuct
+// the volume spec and put it in both the actual and desired state of world. Then the volume direcotory will be cleaned
+// up by reconciler later because the volume entry should be removed by desired state of world's populator if no running
+// container is mounting the volume.
+func (rc *reconciler) reconstructFromDisk(podsDir string) {
+	podVolumes, err := getVolumesFromPodDir(podsDir)
+	if err != nil {
+		glog.Errorf("Cannot get volumes from disk %v", err)
+		return
+	}
+	actualPodVolume := rc.actualStateOfWorld.GetActualPodVolumeSpecNames()
+	desiredPodVolume := rc.desiredStateOfWorld.GetDesiredPodVolumeSpecNames()
+
+	for _, volume := range podVolumes {
+		podName := volume.podName
+		volumeName := volume.volumeSpecName
+		if actualPodVolume[podName][volumeName] || desiredPodVolume[podName][volumeName] {
+			continue
+		}
+
+		//reconstruct Volume object and volumeToMount data structure
+		volumeToMount, err := rc.reconstructVolume(volume)
+		if err != nil {
+			glog.Errorf("Could not construct volume information: %v", err)
+			continue
+		}
+
+		if rc.operationExecutor.IsOperationPending(volumeToMount.VolumeName, volumeToMount.PodName) {
+			continue
+		}
+
+		err = rc.actualStateOfWorld.MarkVolumeAsAttached(
+			volumeToMount.VolumeName, volumeToMount.VolumeSpec, "", volumeToMount.DevicePath)
+		if err != nil {
+			glog.Errorf("Could not add volume information to actual state of world: %v", err)
+			continue
+		}
+		err = rc.actualStateOfWorld.AddPodToVolume(
+			volumeToMount.PodName,
+			types.UID(volumeToMount.PodName),
+			volumeToMount.VolumeName,
+			nil,
+			volumeToMount.OuterVolumeSpecName,
+			volumeToMount.DevicePath)
+		if err != nil {
+			glog.Errorf("Could not add pod to volume information to actual state of world: %v", err)
+			continue
+		}
+		_, err = rc.desiredStateOfWorld.AddPodToVolume(volumeToMount.PodName,
+			volumeToMount.Pod,
+			volumeToMount.VolumeSpec,
+			volumeToMount.OuterVolumeSpecName,
+			volumeToMount.VolumeGidValue)
+		if err != nil {
+			glog.Errorf("Could not add pod to volume information to desired state of world: %v", err)
+			continue
+		}
+	}
+}
+
+func (rc *reconciler) reconstructVolume(volume podVolume) (*operationexecutor.VolumeToMount, error) {
+	volumePluginMgr := rc.desiredStateOfWorld.GetVolumePluginMgr()
+	unescapePluginName := strings.UnescapeQualifiedNameForDisk(volume.pluginName)
```

Comment:
Maybe move this in to `getVolumesFromPodDir()` so that you don't have to worry about it here.